### PR TITLE
Add FindMeetingQuery function

### PR DESCRIPTION
**Pull request by danabreo**

### CODE REVIEW

**Comment by brucezheng** [comment]
What's the difference between this an the other PR?

**Comment by danielabreo** [comment]
> What's the difference between this an the other PR?

#12 should be closed, I incorporated the comments from it, but overall the algorithm is different.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,86 @@
...
+
+    // Check for available time before first 'busy' block.
+    if (TimeRange.START_OF_DAY < allBusyBlocks.get(0).start()) {
+      allAvailableBlocks.add(
+        TimeRange.fromStartEnd(TimeRange.START_OF_DAY, allBusyBlocks.get(0).start(), false)
+      );
+    }
+
+    // Check for available time between middle 'busy' block.
+    for (int i = 0; i < allBusyBlocks.size()-1; ++i) {
```
Comment (brucezheng) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:82]: Nit: Add space before and after ```-```.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,86 @@
...
+      }
+    }
+    allBusyBlocks.add(curBlock);
+    
+    // Create a list of 'available' blocks of time based on 'busy' blocks.
+
+    // Check for available time before first 'busy' block.
+    if (TimeRange.START_OF_DAY < allBusyBlocks.get(0).start()) {
+      allAvailableBlocks.add(
+        TimeRange.fromStartEnd(TimeRange.START_OF_DAY, allBusyBlocks.get(0).start(), false)
```
Comment (brucezheng) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:77]: Nit: No line break between function argument and closing parenthesis ```);```.

**Review by brucezheng — CHANGES_REQUESTED** [review]
I think this implementation is much easier to understand and cleaner. Great job!

For future reference, it's probably better in general to not start a new PR (some exceptions apply). It's good to have history of the revisions so that you can point to some conversations people have had about implementation.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,86 @@
...
+    Set<String> s = event.getAttendees();
+    Collection<String> c = request.getAttendees();
+    for (String person : c) {
+      if (s.contains(person)) return true;
+    }
+    return false;
+  }
+
+  /**
+   * Returns a list of times that are compatible with a given list of events
```
Comment (ccyang314) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:33]: Nit: Summary fragment should be punctuated.

> the fragment is capitalized and punctuated as if it were a complete sentence.

go/java-style#s7.2-summary-fragment

**Review by ccyang314 — APPROVED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,83 @@
...
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> sortedTimes = new ArrayList<TimeRange>();
+    for (Event event : events) {
+      if (sharesAttendees(event, request)) sortedTimes.add(event.getWhen());
+    }
+    Collections.sort(sortedTimes, TimeRange.ORDER_BY_START);
+
+    List<TimeRange> allAvailableBlocks = new ArrayList<TimeRange>();
+    
+    /** Return empty list for an invalid request. */
```
Comment (brucezheng) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:48]: So, javadoc comments are for functions and some other specific things. These are generic implementation comments which should be in the format

```// COMMENT``` or ```/* COMMENT */```.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,83 @@
...
-import java.util.Collection;
+import java.util.*;
+import java.util.stream.Collectors;
 
 public final class FindMeetingQuery {
+
+  /** Checks if the participants of the request are also attending the given event. */
+  private static boolean sharesAttendees(Event event, MeetingRequest request) {
+    Set<String> s = event.getAttendees();
+    Collection<String> c = request.getAttendees();
```
Comment (brucezheng) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:25]: Let's use more descriptive names here.

**Review by brucezheng — CHANGES_REQUESTED** [review]

**Pull request merged**